File: solutions/beecrowd/1032/1032.py
import sys
from math import sqrt

sys.setrecursionlimit(3505)

primes = {
    2: 3,
    3: 5,
    5: 7,
    7: 11,
    11: 13,
    13: 17,
    17: 19,
    19: 23,
    23: 29,
    29: 31,
    31: 37,
    37: 41,
    41: 43,
    43: 47,
    47: 53,
    53: 59,
    59: 61,
    61: 67,
    67: 71,
    71: 73,
    73: 79,
    79: 83,
    83: 89,
    89: 97,
    97: 101,
}


def is_prime(number):
    if number == 2:
        return True
    if number <= 1 or number % 2 == 0:
        return False
    for divisor in range(3, int(sqrt(number)) + 1, 2):
        if number % divisor == 0:
            return False
    return True


def next_prime(number):
    if number in primes:
        return primes[number]

    next_number = number + 1
    while not is_prime(next_number):
        next_number += 1
    primes[number] = next_number

    return primes[number]


def josephus(n, k=2) -> int:
    if n == 1:
        return 1
    return ((josephus(n - 1, next_prime(k)) + k - 1) % n) + 1


for line in sys.stdin:
    n = int(line)

    if n == 0:
        break

    print(josephus(n))

File: solutions/beecrowd/1032/1032.clj
(ns main)

(defn divisible? [number divisor]
  (zero? (rem number divisor)))

(def prime?
  (memoize (fn [number]
             (cond (= number 2) true
                   (or (< number 2) (divisible? number 2)) false
                   :else (loop [divisor 3]
                           (cond
                             (> divisor (Math/sqrt number)) true
                             (divisible? number divisor) false
                             :else (recur (+ divisor 2))))))))

(def next-prime
  (memoize (fn [number]
             (loop [next-number (inc number)]
               (if (prime? next-number)
                 next-number
                 (recur (inc next-number)))))))

;; (def josephus
;;   (memoize (fn [n k]
;;              (if (= n 1)
;;                1
;;                (-> (josephus (dec n) (next-prime k))
;;                    (+ k)
;;                    (dec)
;;                    (mod n)
;;                    (inc))))))

(defn josephus [n]
  (let [primes (take n (iterate next-prime 2))]
    (loop [current-n 1
           result 1
           prime-index 0]
      (if (> current-n n)
        result
        (let [k (nth primes prime-index)
              next-result (inc (mod (+ result k -1) current-n))]
          (recur (inc current-n) next-result (inc prime-index)))))))

(defn main []
  (loop [number (Integer/parseInt (read-line))]
    (when (not= number 0)
      (println (josephus number))
      (recur (Integer/parseInt (read-line))))))

(main)
